samples/benchmarks/coremark/README.rst

```diff
@@ -0,0 +1,170 @@
+
+Overview
+********
+
+The sample runs the CoreMark benchmark on the target CPU.
+CoreMark evaluates the CPU efficiency by performing different algorithms, such as state machine, CRC calculation, matrix manipulation and list processing (find and sort).
+
+To run CoreMark benchmark on the preferred core, press the button that corresponds to the core.
+When the benchmark finishes running, it can be started again by pressing the button assigned to the given core.
+Button and led assignnment is done in devicetree.
+
+Enable ``COREMARK_MODE_FLASH_AND_RUN`` if you want the sample to run for the first time upon startup.
+
+User interface
+**************
+
+Each target CPU has an assigned button and LED that are responsible for triggering the benchmark's start and ``test in progress``.
+
+Default configuration for nRF5340 DK is following:
+
+Button 1:
+   Start the benchmark run on the Application Core.
+Button 2:
+   Start the benchmark run on the Network/Radio Core.
+LED 1:
+   Indicates ``test in progress`` on the Application Core.
+LED 2:
+   Indicates ``test in progress`` on the Network/Radio Core.
+
+
+Configuration
+*************
+
+CoreMark runs tests multiple times. The number of iterations is defined with the ``CONFIG_COREMARK_ITERATIONS`` option.
+By default, the iteration quantity is set to the minimum time required for valid results.
+For valid results, the test should be executed for at least 10 seconds.
+
+CoreMark has three predefined run types, with the predefined data sizes, that are used for data algorithms:
+
+* ``CONFIG_COREMARK_RUN_TYPE_PERFORMANCE`` - defines ``CONFIG_COREMARK_DATA_SIZE`` to 2000 bytes.
+* ``CONFIG_COREMARK_RUN_TYPE_PROFILE`` - defines ``CONFIG_COREMARK_DATA_SIZE`` to 1200 bytes.
+* ``CONFIG_COREMARK_RUN_TYPE_VALIDATION`` - defines ``CONFIG_COREMARK_DATA_SIZE`` to 500 bytes.
+
+However, you can specify a custom ``CONFIG_COREMARK_DATA_SIZE`` value.
+If you want to register your results, see the Submitting Results chapter in the `CoreMark GitHub`_ repository.
+
+Additionally, the ``CONFIG_COREMARK_DATA_SIZE`` option requires available memory space.
+By default, the ``CONFIG_COREMARK_MEMORY_METHOD_STACK`` option is selected.
+In this case, the main thread stack is used for data allocation.
+To change the main stack size, use the ``CONFIG_MAIN_STACK_SIZE`` option.
+
+If you select the ``CONFIG_COREMARK_MEMORY_METHOD_MALLOC`` option, Zephyr :c:func:`k_malloc\k_free` functions will be used to allocate the memory in the heap.
+Adjust the heap size with the ``CONFIG_HEAP_MEM_POOL_SIZE`` config option.
+
+If you select the ``CONFIG_COREMARK_MEMORY_METHOD_STATIC`` option, data will be allocated to the RAM as a regular static variable.
+In this case you do not need to be aware of the main thread stack size or heap memory pool size.
+
+CoreMark can be executed in a multithread mode.
+To specify a number of threads, use the ``CONFIG_COREMARK_THREADS_NUMBER`` option.
+You can configure up to four threads.
```

Comment:
Why only 4 max?

Reply:
This is mostly due to a memory usage.
You can setup more if needed and you have enough or RAM.
But each thread requires ~2+K of main stack.  And in my setup around 17K RAM is used for configuration with 4 threads.

Reply:
Ok, but there might be a completely different board with more memory and multiple cores as well.
Moreover, why do we need that "multithreaded" mode here especially in such a limited (up to 4 threads max) implementation? I'd suggest we keep that sample as simple and minimalistic as possible and have all the multithreading fun with CoreMark Pro ported to Zephyr RTOS with help of POSIX layer. That way it will be possible to spawn even 100 threads and get them automatically scheduled on all cores of the SMP cluster. That requires though finalization of dynamic stack support for `pthread_create()`, see https://github.com/zephyrproject-rtos/zephyr/pull/44727.
Naturally removal of multithreading here will considerably simplify the code here.